src/parser/semantic_check.c

```diff
@@ -15850,3 +15692,198 @@ pt_try_remove_order_by (PARSER_CONTEXT * parser, PT_NODE * query)
 	}
     }
 }
+
+
+/*
+ * pt_check_union_is_foldable - decide if union can be folded
+ * 
+ *  return : union foldability
+ *  parser(in) : Parser context.
+ *  union_node(in) : Union node.
+ */
+STATEMENT_SET_FOLD
+pt_check_union_is_foldable (PARSER_CONTEXT * parser, PT_NODE * union_node)
+{
+  PT_NODE *arg1, *arg2;
+  bool arg1_is_null, arg2_is_null;
+  STATEMENT_SET_FOLD fold_as;
+
+  assert (union_node->node_type == PT_UNION || union_node->node_type == PT_INTERSECTION
+	  || union_node->node_type == PT_DIFFERENCE || union_node->node_type == PT_CTE);
+
+  if (union_node->node_type == PT_CTE)
+    {
+      /* A CTE is a union between the non_recursive and recursive parts */
+      return STATEMENT_SET_FOLD_NOTHING;
+    }
+
+  arg1 = union_node->info.query.q.union_.arg1;
+  arg2 = union_node->info.query.q.union_.arg2;
+
+  arg1_is_null = pt_false_where (parser, arg1);
+  arg2_is_null = pt_false_where (parser, arg2);
+
+  if (!arg1_is_null && !arg2_is_null)
+    {
+      /* nothing to fold at this moment. */
+      fold_as = STATEMENT_SET_FOLD_NOTHING;
+    }
+  else if (arg1_is_null && arg2_is_null)
+    {
+      /* fold the entire node as null */
+      fold_as = STATEMENT_SET_FOLD_AS_NULL;
+    }
+  else if (arg1_is_null && !arg2_is_null)
+    {
+      if (union_node->node_type == PT_UNION)
+	{
+	  fold_as = STATEMENT_SET_FOLD_AS_ARG2;
+	}
+      else
+	{
+	  fold_as = STATEMENT_SET_FOLD_AS_NULL;
+	}
+    }
+  else
+    {
+      assert (!arg1_is_null && arg2_is_null);
+
+      if (union_node->node_type == PT_UNION || union_node->node_type == PT_DIFFERENCE)
+	{
+	  fold_as = STATEMENT_SET_FOLD_AS_ARG1;
+	}
+      else
+	{
+	  assert (union_node->node_type == PT_INTERSECTION);
+	  fold_as = STATEMENT_SET_FOLD_AS_NULL;
+	}
+    }
+
+  return fold_as;
+}
+
+/*
+ * pt_fold_union - fold union node following indicated method
+ *
+ *  return : union node folded
+ *  parser (in) : Parser context.
+ *  union_node (in) : Union node.
+ *  fold_as (in) : indicated folding method
+ */
+PT_NODE *
+pt_fold_union (PARSER_CONTEXT * parser, PT_NODE * union_node, STATEMENT_SET_FOLD fold_as)
+{
+  PT_NODE *arg1, *arg2, *new_node, *next;
+  int line, column;
+  const char *alias_print;
+
+  assert (union_node->node_type == PT_UNION || union_node->node_type == PT_INTERSECTION
+	  || union_node->node_type == PT_DIFFERENCE);
+
+  arg1 = union_node->info.query.q.union_.arg1;
+  arg2 = union_node->info.query.q.union_.arg2;
+
+  line = union_node->line_number;
+  column = union_node->column_number;
+  alias_print = union_node->alias_print;
+  next = union_node->next;
+  union_node->next = NULL;
+
+  if (fold_as == STATEMENT_SET_FOLD_AS_NULL)
+    {
+      /* fold the statement set as null, we don't need to fold orderby clause clause because we return null. */
+      DB_VALUE dummy;
+
+      parser_free_tree (parser, union_node);
+
+      union_node = parser_new_node (parser, PT_VALUE);
```

Comment:
I think `union_node` would be `new_node` or to add `new_node = union_node;`.

It would be great if you remove unused `dummy` and fix typo on the comment: `orderby clause clause`. :)